### PR TITLE
adapter: Subscribe with envelope debezium

### LINKED ISSUE #18436: `SUBSCRIBE`: envelope debezium
`ENVELOPE DEBEZIUM (KEY <expr>)`

When specified, the output of the subscribe becomes (mz_timestamp, mz_before, mz_after). There is no diff field. Within each timestamp, the Debezium envelope accumulates updates into a series of diff pairs (i.e., before/after pairs). Diff pairs are emitted directly in mz_before and mz_after.

The type of mz_before and mz_after is a record type matching the row type of the SUBSCRIBE query.

## Example
Using [get started](https://materialize.com/docs/get-started/). View:
```
SELECT item_name, item_category, revenue FROM item_summary_5min LIMIT 5;
 item_id |     item_name      | item_category |  revenue
---------+--------------------+---------------+-----------
    1024 | health spark       | gadgets       | 514461.99
     512 | vision liquid      | widgets       | 158461.92
     513 | department poppy   | doodads       |    220248
    1025 | kangaroo bomber    | gadgets       |  286929.6
    1792 | Linda transmission | widgets       | 211137.28
```

```
-- Normal output (no envelope)
-- For these examples assume this list represents a full set of updates, e.g. +1/-1 is an update, while a lone -1 is a delete.
COPY (
    SUBSCRIBE (
        SELECT item_id, item_name, item_category, revenue
        FROM item_summary_5min
)) TO STDOUT;

1680012250628	1	706	relatives anethesiologist	clearance	493301.48
1680012251000	1	313	title kite	widgets	310590
1680012251000	-1	313	title kite	widgets	310037.84
1680012251000	-1	44	jumbo Malaysia	doodads	52358.4
1680012251000	-1	899	case hardhat	gadgets	343678.8
1680012251000	1	899	case hardhat	gadgets	345165.3
```

```
COPY (
    SUBSCRIBE (
        SELECT item_name, item_category, revenue
        FROM item_summary_5min
) ENVELOPE DEBEZIUM (KEY item_id)) TO STDOUT;

-- mz_timestamp mz_before                              mz_after
1680012250628   NULL                                    relatives anethesiologist   clearance   493301.48
1680012251000   title kite      widgets     310590      title kite                  widgets    310037.84
1680012251000   jumbo Malaysia  doodads     52358.44    NULL
1680012251000   case hardhat    gadgets     345165.8    case hardhat                gadgets     345165.3
```

```
COPY (
    SUBSCRIBE (
        SELECT item_name, item_category, revenue
        FROM item_summary_5min
) ENVELOPE DEBEZIUM (KEY item_id)) WITH (progress = true) TO STDOUT;

-- mz_timestamp progress    mz_before                               mz_after
1680012250628   F           NULL                                    relatives anethesiologist   clearance   493301.48
1680012251000   F           title kite      widgets     310590      title kite                  widgets    310037.84
1680012251000   F           jumbo Malaysia  doodads     52358.44    NULL
1680012251000   F           case hardhat    gadgets     345165.8    case hardhat                gadgets     345165.3
1680012251001   T           NULL                                    NULL
```

### PR DESCRIPTION
**Pull request by andrewrodriguez-m**
Fixes #18436, progress on #10593 

### Motivation
Outlined in [design doc](https://github.com/MaterializeInc/materialize/blob/084591ec95cd6653a34adb88b87b86cf65247eb1/doc/developer/design/20230418_subscribe_output.md)

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/subscribe.rs [hunk 1]
```diff
@@ -154,44 +165,95 @@ impl ActiveSubscribe {
...
+                                    // [(key, value, -1)] => ("delete", key, value, NULL)
+                                    // everything else => ("key_violation", key, NULL, NULL)
+                                    let value_columns = self.arity - order_by_keys.len();
                                     let mut packer = row_buf.packer();
                                     new_rows.push(match &group[..] {
-                                        [(_, row, 1)] | [(_, _, -1), (_, row, 1)] => {
-                                            // upsert
+                                        [(_, row, 1)] => {
+                                            packer.push(if debezium {
+                                                Datum::String("insert")
```
Comment (mjibson) [inline src/adapter/src/subscribe.rs:177]: Can these both be either consts or hard coded. Don't care which, but should be the same.

File: src/adapter/src/subscribe.rs [hunk 2]
```diff
@@ -154,44 +165,95 @@ impl ActiveSubscribe {
...
+                                                }
+                                            }
                                             for idx in 0..self.arity {
                                                 if !order_by_keys.iter().any(|co| co.column == idx)
                                                 {
                                                     packer.push(datums[idx]);
                                                 }
                                             }
-                                            (start.0, row_buf.clone(), 1)
+                                            (start.0, row_buf.clone(), 0)
```
Comment (mjibson) [inline src/adapter/src/subscribe.rs:196]: Why did all the diffs get changed to 0?

Comment (andrewrodriguez-m) [reply src/adapter/src/subscribe.rs:196]: They don't get read below when outputting them so I tried not to make them meaningful

File: src/adapter/src/subscribe.rs (hunk 1 above)
Comment (andrewrodriguez-m) [reply src/adapter/src/subscribe.rs:177]: done, all hard coded

**Comment by andrewrodriguez-m** [comment]
yeah, i can make prefix all the "after" values with `after_`

File: src/adapter/src/subscribe.rs (hunk 2 above)
Comment (mjibson) [reply src/adapter/src/subscribe.rs:195]: If they sometimes don't get used, this should be an `Option`!

**Review by mjibson — APPROVED** [review]
LGTM. Check with testing if they would like a pass.

File: doc/user/content/sql/subscribe.md [hunk 3]
```diff
@@ -385,6 +386,96 @@ structure:
...
+The output columns are reordered so that all the key columns come before the
+value columns. There are two copies of the value columns, the first is prefixed with
+`before_` and represents the value of the values before this upsert. The current values
+are all prefixed with `after_`.
+
+* Using this modifier, the output rows will have the following
+structure:
+
+   ```sql
+   SUBSCRIBE mview ENVELOPE UPSERT (KEY (key));
```
Comment (philip-stoev) [inline doc/user/content/sql/subscribe.md:405]: Example query says `UPSERT` in the `DEBEZIUM` section of the manual.

File: test/testdrive/subscribe-output.td [hunk 4]
```diff
@@ -268,3 +268,121 @@ DELETE FROM t2 WHERE key = 1
...
+
+> COMMIT
+
+# make sure progress messages are ok
+> BEGIN
+> DECLARE c CURSOR FOR SUBSCRIBE (SELECT * FROM t2) WITH (PROGRESS, SNAPSHOT) ENVELOPE DEBEZIUM (KEY (key))
+> FETCH 1 c
+<TIMESTAMP> true <null> <null> <null> <null>
+
+> FETCH ALL c
```
Comment (philip-stoev) [inline test/testdrive/subscribe-output.td:381]: This unfortunately is timing sensitive, sometimes `FETCH ALL` will only return 1 row, if the second timestamp is not yet available , please use `FETCH 2` instead, see https://github.com/MaterializeInc/materialize/pull/19107

**Review by philip-stoev — APPROVED** [review]
Thank you for the test. I experimented manually with the feature and it seems to work as advertised.

I have annotated all the code lines that are not covered .

File: src/sql/src/plan/error.rs [hunk 5]
```diff
@@ -275,6 +276,9 @@ impl PlanError {
             Self::InvalidKeysInSubscribeEnvelopeUpsert => {
                 Some("All keys must be columns on the underlying relation.".into())
             }
+            Self::InvalidKeysInSubscribeEnvelopeDebezium => {
+                Some("All keys must be columns on the underlying relation.".into())
```
Comment (philip-stoev) [inline src/sql/src/plan/error.rs:280]: This is not covered, as it is distinct from the similar message bug in the context of an UPSERT.

File: test/testdrive/subscribe-output.td (hunk 4 above)
Comment (andrewrodriguez-m) [reply test/testdrive/subscribe-output.td:381]: ah thanks, fixed everywhere

File: src/sql/src/plan/error.rs (hunk 5 above)
Comment (andrewrodriguez-m) [reply src/sql/src/plan/error.rs:280]: Yeah, similar to the EnvelopeUpsert I don't think it's possible to hit this code path today but `panic`ing or ignoring the error if it comes up seems worse than introducing this error path

**Pull request merged**